### ISSUE TITLE
GPIO: Interrupt fails to work after resume from deepsleep @ARC

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#1485]
**_Reported by Qiu Peiyang:_**

Configuration
==========
<Arch> arc
<Board> Quark Se Devboard
<Kernel> unified

Bug detailed description
===================
GPIO interrupt fails to be triggered after ARC resume from deep sleep

Testcase to Reproduce
==================
1. Apply PM framework for driver case, https://gerrit.zephyrproject.org/r/#/c/11215/
2. Apply GPIO deep sleep case, https://gerrit.zephyrproject.org/r/#/c/11262/
3. Connect GPIO_SS_AIN_12 and GPIO_SS_AIN_13
4. comments GPIO Level interrupt case in tests/drivers/gpio/gpio_deepsleep/src/test_gpio_callback.c +103, like this:
/* 
if (test_task(GPIO_INT_LEVEL|GPIO_INT_ACTIVE_LOW))
{ TC_PRINT("Fail to test gpio_callback EDGE_HIGH\n"); return TC_FAIL; }

if (test_task(GPIO_INT_LEVEL|GPIO_INT_ACTIVE_HIGH)) { TC_PRINT("Fail to test gpio_callback EDGE_HIGHn"); return TC_FAIL; }
*/ 
5. make BOARD=quark_se_c1000_ss_devboard flash
6. Connect GPIO_16 from low to high and then from high to low, then this application will run.

Expected Result
=============
GPIO interrupt should be triggerred @ARC side both before deep sleep and after deep sleep.

Actual Result
==========
GPIO interrupt fails to be triggered after ARC resume from deep sleep.

(Imported from Jira ZEP-1624)

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

Maybe related to GH-1003. Will close this issue if GH-1003 can fix it.

**Comment by nashif** [comment]
**_by Sharron LIU:_**

_Qiu Peiyang_ , pending on your inputs to assign.

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

GH-1003 has been merged, but GPIO interrupt still cannot work on ARC side after system resuming from DEEP_SLEEP.
Besides, GPIO read/write can work well after system resuming from DEEP_SLEEP. So I suspect this might be an Interrupt issue.
_Chuck Jordan_  Can you please take a look.

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

QMSI has been updated to v1.4 RC2 and I also update the bootloader for Quark Se at the same time.
So far, other devices can work well after resuming from Deep Sleep including RTC and ADC.
However GPIO interrupt still fails to work after resuming from Deep Sleep and GPIO read/write still works well too.

**Comment by nashif** [comment]
**_by Sharron LIU:_**

_Anselmo L. S. Melo_ , could you find someone take a look?
Per Peiyang's comment, the bootloader has been updated to v1.4, and other devices works fine at ARC.

**Comment by nashif** [comment]
**_by Anselmo L. S. Melo:_**

_Andre Guedes_ , _Iván Briano_ , please take a look.

**Comment by nashif** [comment]
**_by Iván Briano:_**

The attached sample doesn't build.

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

I'm moving this bug to the To Do state because it's a High bug which needs to be fixed in order to release 1.7.

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

* [test_gpio_deepsleep.tgz](https://s3.amazonaws.com/logs.zephyrproject.org/jira/ZEP-1624/test_gpio_deepsleep.tgz) (size: 6242)

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

Upload new test case. Verify it can reproduce this bug.

**Comment by nashif** [comment]
**_by Andre Guedes:_**

HI _Qiu Peiyang_ , could you please clarify what application is running on x86 core?

**Comment by nashif** [comment]
**_by Andre Guedes:_**

I took a look at the test code. In order to work as is, it requires a "dummy" app running on the x86 core. All this dummy app does is: start arc, save x86 context and halt. I utilized the examples/quark_se/start_arc app from QMSI repository to test the code attached to this Jira issue, and it seems to work.

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

_Andre Guedes_  Yes, you are right. x86 code does need to run a "dummy" app.
Weird, I still cannot pass this case. And please try the following steps:

1. Apply PM framework for driver case, https://gerrit.zephyrproject.org/r/#/c/11215/
2. Apply GPIO deep sleep case,  https://gerrit.zephyrproject.org/r/#/c/11262/
3. Connect GPIO_SS_AIN_12 and GPIO_SS_AIN_13
4. comments GPIO Level interrupt case in tests/drivers/gpio/gpio_deepsleep/src/test_gpio_callback.c +103, like this:
/*                                                                                                                                                                                                                              
        if (test_task(GPIO_INT_LEVEL|GPIO_INT_ACTIVE_LOW)) {
                TC_PRINT("Fail to test gpio_callback EDGE_HIGH\n");
                return TC_FAIL;
        }
        
        if (test_task(GPIO_INT_LEVEL|GPIO_INT_ACTIVE_HIGH)) {
                TC_PRINT("Fail to test gpio_callback EDGE_HIGH\n");
                return TC_FAIL;
        }
*/ 
5. make BOARD=quark_se_c1000_ss_devboard flash
6. Connect GPIO_16 from low to high and then from high to low, then this application will run.

Following these steps, I can still reproduce this bug.

**Comment by nashif** [comment]
**_by Andre Guedes:_**

* [andre-output.txt](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-1485/andre-output.txt) (size: 2177)

**Comment by nashif** [comment]
**_by Andre Guedes:_**

So I've done the steps above and the test case is passing. I attached the console output so you can take a look.  [^andre-output.txt] .

Just to remember, I'm running the examples/quark_se/start_arc application from QMSI repository in the x86 core. And just to double check, my gpio wiring setup is: pins 18 and 20 are connected, and pin 30 is connected to GND. To start the test, I connect pin 30 to VCC. All pins belongs to J14 header.

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

_Andre Guedes_ 
We used to use a dummy app on X86, which just enables CONFIG_ARC_INIT=y and then in main function, it calls k_sleep(1) in a while loop. But this dummy app could not work in this case.

Then I try to modify the app at tests/power/multicore/lmt, which enables the power feature on x86 core. I delete the RTC code and make the x86 core sleep without wake up source, and then flash it to x86, but it still fails.
Is there anything wrong with this idea? Or do I need to do any other special settings ?

In my consideration, this case is testing the GPIO functionality @ARC after resuming from deep sleep, our dummy code on x86 should not need to do special settings and since the ARC core can be waken up by RTC alarm, and GPIO read/write can also be verified work, it doesn't make sense GPIO interrupt cannot work.

From our perspective, we also cannot use the code from QMSI repository, as it doesn't belong to zephyr and doesn't comply with zephyr code standard.

I attach the dummy code we are using now. If it's not appropriate, please give some suggestions. If you can provide a dummy app against zephyr, that would be excellent.

Thanks.

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

* [dummy.tgz](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-1485/dummy.tgz) (size: 1266)

**Comment by nashif** [comment]
**_by Iván Briano:_**

If you are going to sleep or deep sleep, as opposed to any of the other power states that don't turn power off, then you need to have an application on the x86 core that makes sure to set everything correctly for wake up, otherwise on wake, the bootloader will not know that it's a wakeup and it will continue booting normally. The ARC appears to be restoring properly because the x86 does the normal call to initialize it and the ARC will restore its context, but a part of the setup is lost. The bootloader, when it doesn't know that it's coming back from sleep (only the x86 sets that up), will do a lot of initialization stuff that would otherwise be avoided, and among them is setting all the interrupt routing masks to their defaults, which is that all interrupts are masked. That's why GPIO read/write work, the ARC will restore the port configuration as part of its waking process, but the interrupt routing registers are not part of that, as they are sticky and we assume they remain configured across the sleep/wakeup boundary.

In short, you can't just have a dummy app with ARC_INIT=y on the x86 side, it has to be an application that sets itself to wake up from sleep and waits there doing nothing, just like that start_arc from QMSI does. If you don't want to use that one, then you'll need to write a similar one for Zephyr.

**Comment by nashif** [comment]
**_by Anselmo L. S. Melo:_**

_Qiu Peiyang_ , any update on this? Thanks.

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

_Iván Briano_ _Anselmo L. S. Melo_ _Andre Guedes_ 
I tried to write a dummy app for zephyr, but didn't make it.
Can you please provide an app and upload to zephyr ? We may need to use it in future cases.

**Comment by nashif** [comment]
**_by Anselmo L. S. Melo:_**

bq. From our perspective, we also cannot use the code from QMSI repository, as it doesn't belong to zephyr and doesn't comply with zephyr code standard.

_Sharron LIU_ , this QMSI code is the same one that was used during power management verification.

**Comment by nashif** [comment]
**_by Anselmo L. S. Melo:_**

_Qiu Peiyang_ , updates?

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

Just get the start_arc.bin and start to verify this issue.
But we still need the start_arc source code for automation, which can use "make flash" to build and flash directly.

**Comment by nashif** [comment]
**_by Anselmo L. S. Melo:_**

https://github.com/quark-mcu/qmsi/tree/master/examples/quark_se/start_arc

This link points to the QMSI 1.3 version - this file is still the same in QMSI 1.4 according to [PR20|https://github.com/quark-mcu/qmsi/pull/20/files].

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

I tried two start_arc apps, one from Sharron and one compiled by myself, both won't resume from deep sleep.
Please help check whether I use the wrong command.
Commands to compile the start_arc:
make SOC=quark_se TARGET=x86 ENABLE_RESTORE_CONTEXT=1

write: 0
read: 0
write: 1
read: 1
write: 1
read: 1
write: 0
read: 0
write: 0
read: 0
write: 0
read: 0
write: 1
read: 1
write: 1
read: 1
write: 0
read: 0
write: 0
read: 0
write: 1
read: 1
write: 1
read: 1
write: 1
read: 1
callback triggered: 1
check enabled callback
callback triggered: 1
check enabled callback
GPIO works well before deep sleep
Power management on device: GPIO_0
PM state to test: 2


Deep sleep entry!
Target device [GPIO_0]: to be suspended
test_gpio_suspend

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

Verify with new case on both lmt and arc, the issue is gone now.
Looks like the app on x86 is the root cause of the issue.
New case is based on zephyr/tests/power/multicore.

X86                arc
active            active
active            gpio operations
active            idle
idle                  idle
active            active(waken up by x86)
active            gpio operations

The new case is attached.
Close this bug.

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

* [test_gpio_deepsleep_arc.tgz](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-1485/test_gpio_deepsleep_arc.tgz) (size: 5100)

**Comment by nashif** [comment]
**_by Qiu Peiyang:_**

New case is test_gpio_deepsleep_arc.tgz.
Will upstream this case later using ZTEST.
